### ISSUE TITLE
Awkward Dynamic Import Treeshaking Output

### ISSUE THREAD

**Issue by lukeed** [rollup/rollup#3451]
katchow!

- Rollup Version: 2.1.0
- Operating System (or Browser): All
- Node Version (if applicable): 10.13.0, 12.x
- Link to reproduction: https://rollupjs.org/repl/?version=2.1.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QiUyMGFhYSUyQyUyMGJiYiUyMCU3RCUyMGZyb20lMjAnLiUyRmJheiclM0IlNUNuJTVDbmNvbnNvbGUubG9nKCdmb28lMjAnJTIwJTJCJTIwYWFhKCklMjAlMkIlMjBiYmIoKSklM0IlNUNuJTVDbmltcG9ydCgnLiUyRmJhcicpLnRoZW4oY29uc29sZS5sb2cpJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIyYmFyLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QiUyMGFhYSUyQyUyMGJiYiUyMCU3RCUyMGZyb20lMjAnLiUyRmJheiclM0IlNUNuJTVDbmV4cG9ydCUyMGRlZmF1bHQlMjBmdW5jdGlvbiUyMCgpJTIwJTdCJTVDbiU1Q3RyZXR1cm4lMjAnaW5zaWRlJTIwYmFyJyUyMCUyQiUyMCUyMGFhYSgpJTIwJTJCJTIwYmJiKCklM0IlNUNuJTdEJTVDbiUyMiUyQyUyMmlzRW50cnklMjIlM0FmYWxzZSU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJiYXouanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyZXhwb3J0JTIwZnVuY3Rpb24lMjBhYWEoKSUyMCU3QiU1Q24lNUN0cmV0dXJuJTIwJ2FhYSclM0IlNUNuJTdEJTVDbiU1Q25leHBvcnQlMjBmdW5jdGlvbiUyMGJiYigpJTIwJTdCJTVDbiU1Q3RyZXR1cm4lMjAnYmJiJyUzQiU1Q24lN0QlNUNuJTIyJTJDJTIyaXNFbnRyeSUyMiUzQWZhbHNlJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzbSUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCU3RCUyQyUyMmV4YW1wbGUlMjIlM0FudWxsJTdE

### Expected Behavior

I did not expect the _main_ entry file to contain a sole `import '....'` file – or `require('....')` when using CJS format.

Instead, I would have expected the contents of `main-*.js` be entrypoint, inlining the `aaa` and `bbb` functions into both files. I suppose I assumed there was some kind of size or count heuristic to determine if an import is worth an extra network request – instead it seems like it's a boolean flag that will always share an export if used in more than 1 file... see [here](https://rollupjs.org/repl/?version=2.1.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QiUyMGFhYSUyMCU3RCUyMGZyb20lMjAnLiUyRmJheiclM0IlNUNuJTVDbmNvbnNvbGUubG9nKCdmb28lMjAnJTIwJTJCJTIwYWFhKCklMjAlMkIlMjBiYmIoKSklM0IlNUNuJTVDbmltcG9ydCgnLiUyRmJhcicpLnRoZW4oY29uc29sZS5sb2cpJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIyYmFyLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QiUyMGFhYSUyMCU3RCUyMGZyb20lMjAnLiUyRmJheiclM0IlNUNuJTVDbmV4cG9ydCUyMGRlZmF1bHQlMjBmdW5jdGlvbiUyMCgpJTIwJTdCJTVDbiU1Q3RyZXR1cm4lMjAnaW5zaWRlJTIwYmFyJyUyMCUyQiUyMCUyMGFhYSgpJTIwJTJCJTIwYmJiKCklM0IlNUNuJTdEJTVDbiUyMiUyQyUyMmlzRW50cnklMjIlM0FmYWxzZSU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJiYXouanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyZXhwb3J0JTIwZnVuY3Rpb24lMjBhYWEoKSUyMCU3QiU1Q24lNUN0cmV0dXJuJTIwJ2FhYSclM0IlNUNuJTdEJTVDbiU1Q25leHBvcnQlMjBmdW5jdGlvbiUyMGJiYigpJTIwJTdCJTVDbiU1Q3RyZXR1cm4lMjAnYmJiJyUzQiU1Q24lN0QlNUNuJTIyJTJDJTIyaXNFbnRyeSUyMiUzQWZhbHNlJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzbSUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCU3RCUyQyUyMmV4YW1wbGUlMjIlM0FudWxsJTdE) for an even simpler REPL.

### Actual Behavior

Any export that is shared across 2+ files (with 1 of those files being the main entry) will automatically force the output's main entry to be a waste of a network request.

**Comment by lukeed** [comment]
Oh. I can double-check in the morning, but I think it's possible to do away with the dummy file entirely (`main.js`) and rely on the `main-1298fa1f.js` file as the main entrypoint for the built application. 

I'm _fairly_ certain that when any `bar-*.js` makes a request to `main-*.js` for its `aaa` export, that browsers will rely on the cached module (pre-parsed) and get the export directly.

I'm not sure how this would behave within Node 13.x but I'd assume it also has a import cache in place too – just like its require() cache.

---

Separately, I've no idea why `main-*.js` exports via:
```js
export { aaa as a };
```

... when `bar-*.js` will just import via:
```js
import { a as aaa } from './main-1298fa1f.js';
```

This seems a bit silly – especially since the source uses the `aaa` identifier only. Since this across multiple file boundaries, Terser wouldn't be able to undo/fix this. IMO leave export/import identifiers alone when auto-code-splitting shared imports.

**Comment by lukastaegert** [comment]
> inlining the aaa and bbb functions into both files

This will quite likely never happen as it is just too dangerous. Duplicating code will break reference equality, which means that e.g. custom attributes added to a function in some way will be broken. However as you realized yourself, this is not the issue that is hitting you.

What is happening is that Rollup tries to maintain the namespace shape of the entry point. Your entry points does not have exports. As Rollup as adding exports to the merged chunk, a "facade" chunk is added that does not have exports. This behaviour is intended for libraries but is indeed irrelevant for web apps. There will be an option soon to configure this. Will be closing this in favour of the tracking issue #3358 .

As for exports being renamed, this is meant to reduce file size for minified builds. Admittedly for very short names it does not make sense. However as you are mentioning terser, with terser it would become something like
```js
import { a as b } from './main-1298fa1f.js';
```

while not renaming it would become more like
```js
import { aaa as b } from './main-1298fa1f.js';
```

What could be improved in Terser is that it could reuse single letter imports as local variables and not try to rename them to other single letter local variables. If you think there is more potential for improvement, please open a separate issue because otherwise it will not be found inside this one.

**Comment by lukastaegert** [comment]
BTW 

> I'm fairly certain that when any bar-*.js makes a request to main-*.js for its aaa export, that browsers will rely on the cached module (pre-parsed) and get the export directly

Yes, this is how all JS runtimes work, which is why Rollup is generating the output it does. This was implemented here: #3354

**Comment by lukeed** [comment]
Thank you for the insight and detailed information :) 

For the most part, it makes complete sense. I still don't think the facade layer is relevant at all, even in regards to "shape" maintenance. While unrealistically simple, in this case, I think the input would serve as a better output than the current output. I'm happy to see a tracking issue for this already & will weigh in there. Apologies for not finding it – I promise that I searched for 20 minutes before posting :)
